src/PhotoNim/shapes.nim

```diff
@@ -5,150 +5,172 @@ import std/options
+        t_hit*: float32
+        world_pt*: Point3D
+        surface_pt*: Point2D
         normal*: Normal
 
-    World* = object
-        shapes*: seq[Shape]
 
+proc newWorld*(): World {.inline.} = World(shapes: @[])
 
-proc newSphere*(transf: Transformation): Sphere {.inline.} = result.transf = transf
-    
-proc newPlane*(transf: Transformation): Plane {.inline.} = result.transf = transf
+proc fire_all_rays*(tracer: var ImageTracer, scenary: World, color_map: proc) = 
+    ## Procedure to actually render an image: we will have to give as an input a function that will enable us to set the color of a pixel
+    for row in 0..<tracer.image.height:
+        for col in 0..<tracer.image.width:
+            tracer.image.setPixel(row, col, color_map(tracer, tracer.fire_ray(row, col), scenary, row, col))
 
-proc newHitRecord*(ray: Ray, t: float32, hit_point: Point3D, uv: Point2D, norm: Normal): HitRecord {.inline.} =
-    HitRecord(ray: ray, t: t, world_point: hit_point, map_pt: uv, normal: norm)
 
-proc areClose*(hit1, hit2: HitRecord): bool {.inline.} = 
-    areClose(hit1.ray, hit2.ray) and areClose(hit1.t, hit2.t) and 
-    areClose(hit1.world_point, hit2.world_point) and  areClose(hit1.map_pt, hit2.map_pt) and 
-    areClose(hit1.normal, hit2.normal) 
+method fastIntersection*(shape: Shape, ray: Ray): bool {.base.} = quit "to overload"
+method rayIntersection*(shape: Shape, ray: Ray): Option[HitRecord] {.base.} = quit "to overload"
 
-proc newWorld*(): World {.inline.} = World(shapes: @[])
+type
+    AABox* = object of Shape
 
+    Sphere* = object of Shape
+        center*: Point3D
+        radius*: float32
 
-method rayIntersection*(shape: Shape, ray: Ray): Option[HitRecord] {.base.} =
-    quit "to overload"
+    Triangle* = object of Shape
+        vertices*: tuple[A, B, C: Point3D]
 
-method fastIntersection*(shape: Shape, ray: Ray): bool {.base.} =
-    quit "to overload"
+    Plane* = object of Shape
 
 
-proc normalOnSphere*(p: Point3D, dir: Vec3f): Normal {.inline.} = 
-    ## Procedure to compute normal on a surface point
-    ## Considering that we are working with an unitary sphere, we can simply use the point coordinates in order to compute the normal. 
-    ## We than just have to chose the direction: we will use the value of the dot product with ray direction as a decisive criterium.
-    - sgn(dot(p.Vec3f, dir)).float32 * newNormal(p.x, p.y, p.z)
+proc newAABox*(min = newPoint3D(0, 0, 0), max = newPoint3D(1, 1, 1), transf = Transformation.id): AABox {.inline.} = AABox(transf: transf, aabb: some(AABB((min, max))))
+proc newSphere*(transf = Transformation.id): Sphere {.inline.} = Sphere(transf: transf)    
+proc newPlane*(transf = Transformation.id): Plane {.inline.} = Plane(transf: transf)
 
 
-proc sphere_uv*(p: Point3D): Point2D = 
-    ## Procedure to compute (u, v) coordinates of the hitpoint
-    var u = arctan2(p.y, p.x) / (2 * PI)
-    if u < 0.0: u += 1.0
+proc uv*[S: Shape](shape: S, pt: Point3D): Point2D = 
```

Comment:
These control statements are at compile time, so it has no overhead.

Reply:
The only problem is that if someone would want to use PhotoNim to write its own shapes should modify the source code instead of writing new code using PhotoNim. Since no one will probably use PhotoNim for that, this approach makes the code less repetitive.